### PR TITLE
[mlir][linalg] Add folder for transpose(transpose) -> transpose

### PR DESCRIPTION
**Pull request by ryanpholt**
Back to back `linalg.transpose` can be rewritten to a single transpose

### CODE REVIEW

**Comment by ryan-holt-1** [comment]
@akshathb

File: mlir/test/Dialect/Linalg/canonicalize.mlir [hunk 1]
```diff
@@ -1051,3 +1051,40 @@ func.func @transpose_identity_perm(%input: tensor<16x32x64xf32>,
...
+  //   CHECK-NOT:   linalg.transpose
+  %transpose1 = linalg.transpose
+      ins(%input:tensor<5x4x3xf32>)
+      outs(%init1:tensor<4x3x5xf32>)
+      permutation = [1, 2, 0]
+  %transpose2 = linalg.transpose
+      ins(%transpose1:tensor<4x3x5xf32>)
+      outs(%init2:tensor<5x4x3xf32>)
+      permutation = [2, 0, 1]
+  func.return %transpose2 : tensor<5x4x3xf32>
```
Comment (cxy-1993) [inline mlir/test/Dialect/Linalg/canonicalize.mlir:1069]: nit: Maybe we should check func.return use %input as result?

File: mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp [hunk 2]
```diff
@@ -1866,6 +1866,36 @@ LogicalResult TransposeOp::fold(FoldAdaptor adaptor,
...
 
+/// Fold transpose with transpose.
+struct FoldTransposeWithTranspose : OpRewritePattern<linalg::TransposeOp> {
+  using OpRewritePattern<linalg::TransposeOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::TransposeOp transposeOp,
+                                PatternRewriter &rewriter) const override {
+    if (auto defTransposeOp =
+            transposeOp.getInput().getDefiningOp<TransposeOp>()) {
+
```
Comment (cxy-1993) [inline mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:1877]: nit: remove redundant blank line

File: mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp [hunk 3]
```diff
@@ -1866,6 +1866,36 @@ LogicalResult TransposeOp::fold(FoldAdaptor adaptor,
...
+      auto defPerms = defTransposeOp.getPermutation();
+      auto perms = transposeOp.getPermutation();
```
Comment (cxy-1993) [inline mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:1879]: nit: expand auto unless it is obvious

**Review by cxy-1993 — CHANGES_REQUESTED** [review]
Thanks for the patch. This patch still needs a few minor changes.

**Comment by ryan-holt-1** [comment]
Thanks for the review @cxy-1993. I have addressed your comments.

**Review by cxy-1993 — APPROVED** [review]
LGTM

File: mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp [hunk 4]
```diff
@@ -1866,6 +1866,35 @@ LogicalResult TransposeOp::fold(FoldAdaptor adaptor,
...
 }
 
+/// Fold transpose with transpose.
+struct FoldTransposeWithTranspose : OpRewritePattern<linalg::TransposeOp> {
+  using OpRewritePattern<linalg::TransposeOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::TransposeOp transposeOp,
+                                PatternRewriter &rewriter) const override {
+    if (auto defTransposeOp =
+            transposeOp.getInput().getDefiningOp<TransposeOp>()) {
```
Comment (ftynse) [inline mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp:1876]: https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

**Review by ftynse — APPROVED** [review]

**Pull request merged**